cdap-master/src/main/java/co/cask/cdap/data/tools/ReplicationStatus.java

```diff
@@ -0,0 +1,290 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * The Replication Status tool is used to track the status of replication from Master to a Slave Cluster.
+ * The tool reads writeTime and replicateTime for all tables and uses it to determine if all CDAP tables have been
+ * completely replicated to the Slave cluster.
+ * */
+public class ReplicationStatus {
+
+  private static int writeTime = 0;
+  private static int replicateTime = 1;
+
+  private static CConfiguration cConf = CConfiguration.create();
+  private static Configuration hConf = HBaseConfiguration.create();
+
+  protected static final Logger LOG = LoggerFactory.getLogger(ReplicationStatus.class);
+
+  private static String masterCluster = null;
+  public static void main (String args[]) throws Exception {
+
+    if (args.length < 1 || args[0].equals("help")) {
+      System.out.printf("Tool to check replication status to Slave Cluster.\n\n");
+      System.out.printf("Run this tool on the Slave cluster as:\n\n");
+      System.out.println("#ReplicationStatus <master cluster id:>");
+      return;
+    } else if (args[0].equals("debugdump")) {
+      dumpReplicationStateTable();
+    } else {
+      masterCluster = args[0];
+      checkHBaseReplicationComplete();
+    }
+  }
+
+  private static void checkHBaseReplicationComplete() throws Exception {
+
+    HBaseTableUtil tableUtil = new HBaseTableUtilFactory(cConf).get();
+    HTable hTable = tableUtil.createHTable(hConf, getReplicationStateTableId(tableUtil));
+    long shutdownTime = 0L;
+
+    // Scan the table to scan for all regions.
+    ScanBuilder scan = getScan(tableUtil);
+    Result result;
+
+    HashMap<String, ArrayList<Long>> regionMap = new HashMap<>();
+    try (ResultScanner resultScanner = hTable.getScanner(scan.build())) {
+      while ((result = resultScanner.next()) != null) {
+        ByteBuffer row = ByteBuffer.wrap(result.getRow());
+        String clusterID = new String(getNextPart(row), "UTF-8");
+        String nextPart = new String(getNextPart(row), "UTF-8");
+
+        if (nextPart.matches(Constants.ReplicationStatusTool.SHUTDOWN_TIME_ROW)) {
+          // get shutdown time from master cluster
+          if (clusterID.matches(masterCluster)) {
+            shutdownTime = getShutdownTime(result);
+          }
+          continue;
+        }
+
+        if (clusterID.matches(masterCluster)) {
+          long writeTime = getTimeFromResult(result, Constants.ReplicationStatusTool.WRITE_TIME);
+          updateWriteTime(regionMap, nextPart, writeTime);
+        } else {
+          long replicateTime = getTimeFromResult(result, Constants.ReplicationStatusTool.REPLICATE_TIME);
+          updateReplicateTime(regionMap, nextPart, replicateTime);
+        }
+      }
+    } catch (Exception e) {
+      LOG.error("Error while reading table:" + e);
+      throw Throwables.propagate(e);
+    } finally {
+      hTable.close();
+    }
+
+    // Make sure master shutdown time is present
```

Comment:
Shutdown time should be read separately from the scan though. There is a condition in the loop that only ever applies to one iteration through the loop.